crates/extensions/tedge_timer_ext/src/builder.rs

```diff
@@ -0,0 +1,202 @@
+use crate::actor::TimerActor;
+use crate::actor::TimerId;
+use crate::SetTimeout;
+use crate::Timeout;
+use async_trait::async_trait;
+use std::collections::HashMap;
+use std::convert::Infallible;
+use std::sync::Arc;
+use tedge_actors::Actor;
+use tedge_actors::Builder;
+use tedge_actors::ChannelError;
+use tedge_actors::DynSender;
+use tedge_actors::Message;
+use tedge_actors::MessageBoxPlug;
+use tedge_actors::MessageBoxSocket;
+use tedge_actors::NoConfig;
+use tedge_actors::Sender;
+use tedge_actors::ServiceMessageBoxBuilder;
+use tokio::sync::Mutex;
+
+pub struct TimerActorBuilder {
+    box_builder: ServiceMessageBoxBuilder<SetTimeout<TimerId>, Timeout<TimerId>>,
+}
+
+impl Default for TimerActorBuilder {
+    fn default() -> Self {
+        TimerActorBuilder {
+            box_builder: ServiceMessageBoxBuilder::new("Timer", 16),
+        }
+    }
+}
+
+impl Builder<(TimerActor, <TimerActor as Actor>::MessageBox)> for TimerActorBuilder {
+    type Error = Infallible;
+
+    fn try_build(self) -> Result<(TimerActor, <TimerActor as Actor>::MessageBox), Self::Error> {
+        Ok(self.build())
+    }
+
+    fn build(self) -> (TimerActor, <TimerActor as Actor>::MessageBox) {
+        let actor = TimerActor::default();
+        let actor_box = self.box_builder.build();
+        (actor, actor_box)
+    }
+}
+
+impl<T: Message> MessageBoxSocket<SetTimeout<T>, Timeout<T>, NoConfig> for TimerActorBuilder {
+    fn connect_with(
+        &mut self,
+        peer: &mut impl MessageBoxPlug<SetTimeout<T>, Timeout<T>>,
+        config: NoConfig,
+    ) {
+        let mut adapter = GenericTimerAdapter::new(peer);
+        self.box_builder.connect_with(&mut adapter, config);
+    }
+}
+
+/// A message adapter used by actors to send timer requests with a generic payload `SetTimeout<T>`
+/// and to receive accordingly timer responses with a generic payload `Timeout<T>`,
+/// while the timer actor only handles `usize` payloads.
+///
+/// This adapter uses a cache from assigned `TimerId` to actual data of type `T`.
+/// This cache is populated for each received `SetTimeout<T>` request
+/// on translation into a `SetTimeout<TimerId>`; and cleaned when the `Timeout<TimerId>` is received
+/// and translated back into a `Timeout<TimerId>`.
+///
+/// Technically, the cache needs to be wrapped behind an `Arc<Mutex< ... >>`,
+/// because being accessed by two independent actors: the caller and the timer actors.
```

Comment:
Why should this internal cache be exposed to the caller? He should only be dealing the `Timeout<T>`, right? Any translation from `TimerId` to `T` should ideally be just an internal logic of the timer actor and not exposed to the peer, right?

Reply:
This is an internal logic. None of these structs are `pub`. Have a look to the tests.